src/components/cl/hier/allreduce/allreduce_split_rail.c

```diff
@@ -0,0 +1,368 @@
+
+    task_ag->n_deps = 1;
+    ucc_schedule_add_task(schedule, task_ag);
+    ucc_event_manager_subscribe(&task_ar->em, UCC_EVENT_COMPLETED, task_ag,
+                                ucc_dependency_handler);
+
+    schedule->super.post     = ucc_schedule_start;
+    schedule->super.progress = NULL;
+    schedule->super.finalize = ucc_cl_hier_allreduce_split_rail_frag_finalize;
+
+    *frag_p = schedule;
+    return status;
+
+err_ag:
+    if (task_ar) {
+        ucc_collective_finalize(&task_ar->super);
+    }
+err_ar:
+    if (task_rs) {
+        ucc_collective_finalize(&task_rs->super);
+    }
+err_rs:
+    ucc_cl_hier_put_schedule(schedule);
+    return status;
+}
+
+static inline void get_n_frags(ucc_base_coll_args_t *coll_args,
+                               ucc_cl_hier_team_t *team, int *n_frags,
+                               int *pipeline_depth)
+{
+    ucc_cl_hier_lib_config_t *cfg     = &UCC_CL_HIER_TEAM_LIB(team)->cfg;
+    size_t                    msgsize = coll_args->args.dst.info.count *
+                     ucc_dt_size(coll_args->args.dst.info.datatype);
+    int min_num_frags;
+
+    *n_frags = 1;
+    if (msgsize > cfg->allreduce_split_rail_frag_thresh) {
+        min_num_frags = msgsize / cfg->allreduce_split_rail_frag_size;
+        *n_frags = ucc_max(min_num_frags, cfg->allreduce_split_rail_n_frags);
+    }
+    *pipeline_depth =
+        ucc_min(*n_frags, cfg->allreduce_split_rail_pipeline_depth);
+}
+
+static ucc_status_t
+ucc_cl_hier_split_rail_allreduce_start(ucc_coll_task_t *task)
+{
+    ucc_schedule_pipelined_t *schedule =
+        ucc_derived_of(task, ucc_schedule_pipelined_t);
+
+    cl_info(task->team->context->lib,
+            "posting split_rail ar, sbuf %p, rbuf %p, count %zd, dt %s, op %s, "
+            "inplace %d, pdepth %d, frags_total %d",
+            task->bargs.args.src.info.buffer, task->bargs.args.dst.info.buffer,
+            task->bargs.args.dst.info.count,
+            ucc_datatype_str(task->bargs.args.src.info.datatype),
+            ucc_reduction_op_str(task->bargs.args.op),
+            UCC_IS_INPLACE(task->bargs.args), schedule->n_frags,
+            schedule->super.n_tasks);
+
+    return ucc_schedule_pipelined_post(task);
+}
+
+UCC_CL_HIER_PROFILE_FUNC(ucc_status_t, ucc_cl_hier_allreduce_split_rail_init,
+                         (coll_args, team, task),
+                         ucc_base_coll_args_t *coll_args, ucc_base_team_t *team,
+                         ucc_coll_task_t **task)
+{
+    ucc_cl_hier_team_t *cl_team = ucc_derived_of(team, ucc_cl_hier_team_t);
+    int                 n_frags, pipeline_depth;
+    ucc_cl_hier_lib_config_t *cfg   = &UCC_CL_HIER_TEAM_LIB(cl_team)->cfg;
+    size_t                    count = coll_args->args.dst.info.count;
+    size_t data_size = count * ucc_dt_size(coll_args->args.dst.info.datatype);
+    ucc_cl_hier_schedule_t *schedule;
+
+    ucc_status_t status;
+
+    if (!ucc_topo_isoppn(team->params.team->topo)) {
```

Comment:
this is not unlikely. This line checks if core team is "isoppn" (suggest better name) - ppn is the same across all the nodes. If check is false the algorithm can not be used. But in general we have no prior knowledge on the core team layout.

Reply:
why we can't check it from cl hier team and disable this algorithm? anyway i think it's unlikely because usually ppn is the same and even if it's not the cost of "unlikely()" is much less than cost of fallback

Reply:
we can check at team create (but user still can force enable via TUNE) - i'll add that. "usually ppn is the same" - this is false assumption. in ML world it is likely true. In HPC world it is quite often false. in general i totally agree with you that this "likely/unlikely" stuff has nearly 0 effect. I consider it more like readability hint for developers.